src/psyclone/psyir/nodes/call.py

```diff
@@ -593,3 +600,172 @@ def _location_txt(node):
             f"Failed to find a Routine named '{rsym.name}' in "
             f"{_location_txt(root_node)}. This is normally because the routine"
             f" is within a CodeBlock.")
+
+    def get_argument_routine_match(self, routine: Routine):
+        """Return a list of integers giving for each argument of the call
```

Comment:
The passing back of a list of integers seems a bit clunky. Could you explain why you need this? Could we pass back the Symbols themselves perhaps?

Reply:
There's an excellent purpose of returning the numbers:
This can be used by the inlining functionality.
Finding the right routine also means that you need to check whether the arguments match.
Once you know this, for inlining, you need to know which call argument is related to which subroutine argument.
This is given by these numbers.
Alternatively, a list "ret_arg_idx:List[int]=None" could be handed over as an argument to this function which is only updated if it's provided:

```
if ret_arg_idx is not None:
    ret_arg_idx[:] = ... # The list with the indices
```

Reply:
As always, looking at the tests helped me think this through a bit more and I realise that my list idea won't work. I think we do want a `dict` but the keys should be the routine's formal arguments (either their names or the Symbols) and the values should be the actual arguments provided to the Call. I think this then mirrors what happens when calling a Python function which is encouraging.

Reply:
This is still outstanding?

Reply:
I think that a list of integers is sufficient. The optional arguments of a subroutine can be handled as follows (future work):

For inlining with optional arguments:
1. create a copy of the routine.
2. Search for optional arguments which are not in the list of indices.
3. For all of the optional arguments, replace PRESENT(...) with booleans.
4. Eliminate branches which are always true/false.
5. Check that there's no unused access to an optional argument left.

That's it. It's an algorithm which might not work in all possible cases, but it should with Nemo from what I saw so far.

Reply:
This is an example how easily this list of indices can be used (it's not yet finished):

https://github.com/stfc/PSyclone/blob/martin_nemo_call_trace/src/psyclone/psyir/transformations/inline_trans.py#L990

Reply:
You're argument about being able to use the list of integers to map in both directions seems good, thanks.